sisl/viz/plotly/plots/bands.py

```diff
@@ -865,3 +888,67 @@ def _plot_Kdiff(self, band1, band2, E=None, offsetE=False):
         plt.update_layout({"title": f"Delta K between bands {band1} and {band2}", 'xaxis_title': 'Delta k', 'yaxis_title': 'Energy [eV]'})
 
         return plt
+
+    def effective_mass(self, band, k, band_spin=0, n_points=10, k_direction="symmetric"):
+        """
+        Calculates the effective mass from the curvature of a band in a given k point.
+
+        NOTE: The calculations done here are only valid if there are no band crossings in the
+        range that is being fitted.
+
+        Parameters
+        -----------
+        band: int
+            The index of the band that we want to fit
+        k: float or str
+            The k value where we want to find the curvature of the band to calculate the effective mass.
+        band_spin: int, optional
```

Comment:
change to `spin`, I think the meaning is implicit. :)

Reply:
Hehe, that's one of the problems that we'll have to address with how I currently have settings implemented.

Right now, the way I let the developer make use of plot settings is the following:

```python
class MyPlot(Plot):

    _parameters = (
        InputField(key="first_setting", ...)
    )

    def some_func(self, first_setting):
        # The first_setting argument here will receive the value of the setting by default,
        # unless another value is explicitly passed
```

After some trial and error on how to use settings, I was quite happy with this approach because it makes methods reusable from the "outside". E.g:

```python
class MyOtherPlot(Plot):

    def do_stuff(self):
        MyPlot.some_func(self, 4)
```

That is, I can still call the `some_func` function of `MyPlot` without needing to have the `first_setting` defined in `MyOtherPlot`.

It also lets `Plot` know which functions make use of `first_setting`, so when the setting is updated the appropiate functions are ran again.

However, this unfortunately leads to not being able to use some words as arguments. For example, in `BandsPlot`, `spin` is a setting. If I name this argument `spin` and no value is provided when the method is called, the value of the `spin` setting will be passed, which will result in an error :(

Reply:
Are you doing:
```python
class MyPlot(Plot):
    ...

class MyOtherPlot(Plot):
    ...
```
or 
```python
class MyPlot(Plot):
    ...

class MyOtherPlot(MyPlot):
    ...
```

The former I really think you should fix ;) Don't use stuff from non-subclassed classes. 

If it was a typo and you are actually doing the latter, then ok. :)

If you want to re-use settings it seems better to have another way of handling this.

Reply:
Wouldn't it be better if the `settings` have a hook (callback) that calls the appropriate class methods?
In that way you can *link* settings to different plots, and if they don't need linking, they get their own setting?

This would also leverage that problem of `band_spin` vs. `spin` vs. `eig_spin` ... (I think)?

Reply:
> If you want to re-use settings it seems better to have another way of handling this.

The point was not to reuse settings, it was to be able to use a function without necessarily having all the settings defined. Silly example:

```
class BandsPlot(Plot):
    _parameters = (
        RangeInput(key="bands_range")
    )

    def draw_bands(self, bands_range):
        ...draw the bands within that range

class Band0Plot(BandsPlot):
    "A plot that draws only the first band"

    def draw_bands(self):
        BandsPlot.draw_bands(self, [0,0])
```

The point being, even though `Bands0Plot` is an extension of `BandsPlot`, it doesn't make sense to allow a `bands_range` value other than `[0,0]`, therefore it shouldn't be a setting. We still should be able to use the method to draw the bands.

> Wouldn't it be better if the settings have a hook (callback) that calls the appropriate class methods?
> In that way you can link settings to different plots, and if they don't need linking, they get their own setting?

Would you declare explicitly which method does each setting need to call when updated? That's what I wanted to avoid to make developement easier.

Reply:
> ```python
> class BandsPlot(Plot):
>     _parameters = (
>         RangeInput(key="bands_range")
>     )
> 
>     def draw_bands(self, bands_range):
>         ...draw the bands within that range
> 
> class Band0Plot(BandsPlot):
>     "A plot that draws only the first band"
> 
>     def draw_bands(self):
>         BandsPlot.draw_bands(self, [0,0])
> ```
> 
> The point being, even though `Bands0Plot` is an extension of `BandsPlot`, it doesn't make sense to allow a `bands_range` value other than `[0,0]`, therefore it shouldn't be a setting. We still should be able to use the method to draw the bands.

But here it would make sense to just have `BandsPlot` inherit from `Bands0Plot`.

> > Wouldn't it be better if the settings have a hook (callback) that calls the appropriate class methods?
> > In that way you can link settings to different plots, and if they don't need linking, they get their own setting?
> 
> Would you declare explicitly which method does each setting need to call when updated? That's what I wanted to avoid to make developement easier.

I still need to wrap my head around all of this (as you may see from my questions).
Since you still have to call a method explicitly, why not have a decorator taking care of this. That could hook in settings etc. for specific methods?

Reply:
> But here it would make sense to just have BandsPlot inherit from Bands0Plot.

Why? `Band0Plot` is more specific than `BandsPlot` so the inheritance should be `BandsPlot` -> `Band0Plot`, no?

> Since you still have to call a method explicitly, why not have a decorator taking care of this. That could hook in settings etc. for specific methods?

Actually I'm using a decorator:

https://github.com/zerothi/sisl/blob/229c04680f41254673999629d349104274282fc8/sisl/viz/plotly/configurable.py#L892-L948

which is applied to every plot (or session) method here:

https://github.com/zerothi/sisl/blob/229c04680f41254673999629d349104274282fc8/sisl/viz/plotly/configurable.py#L405-L407

i.e. when you define a class that inherits from Configurable.

But you mean something more explicit like:

```python
class BandsPlot(Plot):
 
    _parameters = (
        RangeInput(key="bands_range")
    )
    
    @uses("bands_range")
    def some_func(self, bands_range):
        ...
```

right?

Reply:
> > But here it would make sense to just have BandsPlot inherit from Bands0Plot.
> 
> Why? `Band0Plot` is more specific than `BandsPlot` so the inheritance should be `BandsPlot` -> `Band0Plot`, no?
>

I would argue the other way around.
`Band0Plot` is your basic bands plot. Then you add functionality in `BandsPlot`. Hehe.

> But you mean something more explicit like:
> 
> ```python
> class BandsPlot(Plot):
>  
>     _parameters = (
>         RangeInput(key="bands_range")
>     )
>     
>     @uses("bands_range")
>     def some_func(self, bands_range):
>         ...
> ```
> 
> right?

Here you could do something like:
```python
@uses("bands_range")
def some_func(self):
    do whatever
```
and the decorator would add a `kwarg` equivalent to `bands_range`

Reply:
I really need to figure out why you have these settings implemented like this to better get an overview :)

Reply:
> Band0Plot is your basic bands plot. Then you add functionality in BandsPlot. Hehe.

But `Band0Plot` is nothing without a general method that draws bands, which should be implemented in `BandsPlot` :thinking:

> @uses("bands_range")
> def some_func(self, **kwargs):
>     do whatever

```python
@uses("bands_range")
def some_func(self, **kwargs):
    do whatever
```

right?

> I really need to figure out why you have these settings implemented like this to better get an overview :)

Basically the only difference between decorators and the current implementation is that I thought it would be messy if there was decorators all around, and it would be more difficult for begginers to understand and therefore more difficult for them to create plots. But maybe it outweights the possibility of having bugs that are difficult to identify with the current implicit approach.

Reply:
You don't need the above. See e.g. this:
```python
from functools import wraps
def add_argument(*new_args):
    def true_decorator(func):
        @wraps(func)
        def func_add_argument(*args, **kwargs):
            for new_arg in new_args:
                arg = kwargs.pop(new_arg)
                print(new_arg, arg)
                # do something with arg
            return func(*args, **kwargs)
        return func_add_argument
    return true_decorator


@add_argument("temp")
def no_args():
    print("Hi from no_args")

no_args(temp="sthaoe")
```

In any case it is difficult. I think it has to be the simplest and most comprehensible one since that is easier to maintain ;)
But perhaps there are cases where your method is superior. :) I just need to get started on understanding.

Reply:
Aah ok but then if you want to use the `"temp"` setting inside `no_args` you need to get the value of the setting yourself from `self.settings` (`self.settings["temp"]`)?. So you have to rely on the developer not forgetting to say what settings they are using for updates to run properly.

Reply:
Could you perhaps guide me towards one of the simpler classes where this is used so I can learn ;)

Reply:
In a call or here? :thinking:

Reply:
Perhaps we should make a call at some point. For now I really need to concentrate on fixing the documentation up and get plotly integrated with docs.